doc: add pages about data and key storage (#22683)
nrfconnect/sdk-nrf: doc/nrf/app_dev/data_storage.rst

.. figure:: /images/data_storage_rramc.svg
   :alt: Data storage using RRAMC

   Data storage using RRAMC

This option offers the following features:

* Write without erase - RRAMC allows writing and overwriting without requiring erase operations.
* Error correction - Built-in ECC (Error Correction Code) can detect and correct up to two bit errors per 128-bit word line.
* Power management - Automatic standby or power-down modes with configurable latency.
* Region protection - Configurable memory regions with read, write, execute, and secure access controls.
* Write-once support - Optional write-once protection for memory regions.
* Buffered writes - Support for fast buffered writes to contiguous memory regions.
* Power-failure protection - Built-in power-fail comparator to prevent writes during low voltage conditions.

The RRAMC provides several protection mechanisms, including:

* Region protection - Memory can be divided into configurable regions with different access permissions.
* Immutable boot region - Configurable read-only boot region starting at address ``0x00000000``.

> * One-time programmable protection - Optional write-once protection for UICR.

Reviewer: ```suggestion
* One-time programmable protection - Optional write-once protection for UICR.
```
```suggestion
* One-time programmable protection (Hardware-fixed for UICR).
```

RRAMC regions can be configured as write-once (optionally), but UICR is hardware-fixed to be write-once

.. _data_storage:

> .. _memory_storage:

Author: This is just to keep the breadcrumb trace from the old doc bundles. This is how we do it. I will raise this with my peers.

Data storage in the |NCS|
#########################

.. contents::
   :local:
   :depth: 2

The |NCS| offers multiple storage solutions for storing data persistently in non-volatile memory.

This page provides an overview of the available storage alternatives to help you make an informed choice for your application's data storage needs.

For information about storing cryptographic keys, see :ref:`key_storage`.

General recommendations
***********************

For general data storage needs, Nordic Semiconductor recommends starting with the following options:

* NVMC - For applications requiring direct access to non-volatile memory.
* NVS - For most applications requiring simple persistent storage.
* ZMS - For nRF54H and nRF54L Series devices, use ZMS for persistent storage instead of NVS.
* RRAMC and ZMS - For nRF54L Series devices, use RRAMC together with ZMS, instead of NVMC and NVS.
  RRAMC provides the hardware capabilities for RRAM memory management, while ZMS offers an optimized data management system that extends memory lifespan.
* File Systems - For applications requiring file-based data organization.

> * Settings - For configuration data and runtime state management.

Author: Made also edits to NVS, ZMS, Settings sections, in addition to this line.

     - Yes
     - Yes
     - Yes
     - No
   * - :ref:`ZMS for nRF54L15 and nRF54H20 <data_storage_zms>`
     - Yes
     - Yes
     - Yes
     - No
   * - :ref:`File Systems <data_storage_filesystems>`
     - Yes
     - Yes
     - No
     - No
   * - :ref:`Settings <data_storage_settings>`
     - Yes
     - Yes*
     - No
     - No
   * - :ref:`PSA Protected Storage <data_storage_psa_protected>`
     - Yes
     - Yes
     - Yes
     - Yes

\* The Settings subsystem can use different backends. If the backend has integrity, Settings also does.

Feature descriptions
====================

See the following sections for descriptions of the storage features from the table above.

Partitioning
------------

Partitioning prevents storage subsystems from overwriting each other's data by giving them ownership over different non-volatile memory regions.
In the |NCS|, you can use either :ref:`Devicetree fixed flash partitions <zephyr:flash_map_api>` or the :ref:`Partition Manager <partition_manager>` to manage non-volatile memory partitions.

Integrity
---------

Data integrity ensures that the data you read is the same as the data you wrote.

> This is typically implemented using Cyclic Redundancy Check (CRC) or cryptographic hash functions like SHA-2.

Author: It was just an example in the blog post. But indeed, it can be read as something that is being implemented in the NCS. Changed to SHA-256.

Isolation
---------

Isolation provides security separation between the Non-Secure Processing Environment (NSPE) and Secure Processing Environment (SPE) as part of the Arm Platform Security Architecture (PSA).
This feature is available when using :ref:`Trusted Firmware-M (TF-M) <ug_tfm>`.

Encryption
----------

Encryption provides additional data-at-rest protection by encrypting stored data.
This adds confidentiality to prevent unauthorized access to sensitive information stored in non-volatile memory.

.. _data_storage_alternatives:

> Storage alternatives

Author: Maybe in its secondary or colloquial meaning. The primary meaning is synonymous with "option" or "choice".

********************

The |NCS| provides several storage mechanisms for storing data persistently in non-volatile memory.
Each alternative offers different features and trade-offs in terms of security, performance, and ease of use.

.. _data_storage_nvmc:

NVMC
====

.. list-table:: NVMC features
   :widths: auto
   :header-rows: 1

   * - Feature
     - Partitioning
     - Integrity
     - Isolation
     - Encryption
   * - **NVMC**
     - No
     - No
     - No
     - No

The Non-Volatile Memory Controller (NVMC) provides the most basic method for writing directly to non-volatile memory.
The |NCS| provides the `NVMC driver through nrfx <nrfx NVMC driver_>`_ as a helper for this functionality.

.. note::
   Nordic Semiconductor does not recommend this method for persistent storage, as it places significant responsibility on the developer.
   For simple data storage needs, consider using :ref:`NVS <data_storage_nvs>` instead.

> .. figure:: /images/data_storage_nvmc.svg

Author: I'm blind 🙈

* Automatic integrity - Uses CRC to verify data integrity automatically.
* Wear leveling - Distributes writes across flash sectors to extend memory lifetime.
* Atomic operations - Ensures data consistency during power loss.

NVS automatically manages metadata for stored elements, including the following components:

* Element ID
* Data offset in the sector
* Data length
* CRC for integrity checking

During initialization, NVS verifies all stored data and ignores any data with missing or incorrect metadata.

Configuration
-------------

To enable NVS, set the following Kconfig options:

* :kconfig:option:`CONFIG_NVS`
* :kconfig:option:`CONFIG_FLASH`
* :kconfig:option:`CONFIG_FLASH_MAP`

Usage example
-------------

See Zephyr's :zephyr:code-sample:`nvs` sample for an example of how to use NVS.

.. _data_storage_rramc:

RRAMC
=====

.. note::
   RRAMC is only available on the `nRF54L15 SoC <nRF54L15 RRAMC_>`_ and should be used with :ref:`ZMS <data_storage_zms>`.

.. list-table:: RRAMC features
   :widths: auto
   :header-rows: 1

   * - Feature
     - Partitioning
     - Integrity
     - Isolation
     - Encryption
   * - **RRAMC**
     - Yes
     - Yes

>      - Yes

Author: I've misinterpreted "region protection" as isolation https://docs.nordicsemi.com/bundle/ps_nrf54L15/page/rramc.html#ariaid-title6 . I need yours and @rghaddab help here.

>      - Yes

Reviewer 2: @greg-fer Isolation is not related to ZMS, but implemented by upper layers, So I don't understand what that means for ZMS.
@tomi-font Am I missing something here ?

>      - Yes

Author: Should I then say that Isolation is "inherited" for ZMS and RRAMC when TF-M is being used? Like we do with integrity for Security, that inherits it from the backend? @tomi-font , @rghaddab

* Error correction - Built-in ECC (Error Correction Code) can detect and correct up to two bit errors per 128-bit word line.
* Power management - Automatic standby or power-down modes with configurable latency.
* Region protection - Configurable memory regions with read, write, execute, and secure access controls.
* Write-once support - Optional write-once protection for memory regions.
* Buffered writes - Support for fast buffered writes to contiguous memory regions.
* Power-failure protection - Built-in power-fail comparator to prevent writes during low voltage conditions.

The RRAMC provides several protection mechanisms, including:

* Region protection - Memory can be divided into configurable regions with different access permissions.
* Immutable boot region - Configurable read-only boot region starting at address ``0x00000000``.
* One-time programmable protection - Hardware-fixed to write-once for UICR.
  Other regions can optionally be configured to be write-once.
* Erase protection - Ability to block ``ERASEALL`` operations.

For more information, see `nRF54L15 SoC datasheet <nRF54L15 RRAMC_>`_.

.. _data_storage_zms:

ZMS
===

.. note::

>    ZMS is only available on the nRF54L and nRF54H Series devices.

Author: Changed "only" to "recommended" here and in the General recommendations above.

>    For nRF54L Series devices, ZMS should be used with :ref:`RRAMC <data_storage_rramc>` instead of NVMC and NVS.

Reviewer 2: ZMS is using Flah API.
Depending on which driver is enabled for a platform, the Flash API will use the read/write calls of the driver API.
So ZMS will automatically use whatever is defined by the user in the device tree. This has nothing to do with ZMS

It is available to both NSPE and SPE, but data stored from one environment is not accessible from the other.

When to use PSA Protected Storage
---------------------------------

Consider using PSA Protected Storage in the following cases:

* You need data encrypted at rest.
* You require PSA Certified compliant storage.
* You need write-once functionality for immutable data.
* You are building security-critical applications.

When comparing PSA Protected Storage with :ref:`data_storage_nvs`, consider the following:

* Choose PSA Protected Storage when you need:

  * Encrypted data storage
  * Security isolation between environments
  * PSA Certified compliance
  * Write-once functionality

* Choose NVS when you need:

  * Simple key-value storage
  * Lower overhead
  * Broader platform compatibility
  * No security isolation requirements

Configuration
-------------

PSA Protected Storage requires :ref:`TF-M <ug_tfm>` to be enabled.
The necessary Kconfig options are typically enabled automatically when TF-M is configured.

Usage example
-------------

> For a usage example, see Zephyr's :zephyr:code-sample:`psa_protected_storage` sample and the :ref:`tfm_secure_peripheral_partition` sample in the |NCS|.

Author: Wrong assumption based on https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/security/tfm/tfm_architecture.html#application_rot_services

.. _data_storage:
.. _memory_storage:

Data storage in the |NCS|
#########################

.. contents::
   :local:
   :depth: 2

The |NCS| offers multiple solutions for storing data persistently in non-volatile memory.

This page provides an overview of the available storage solutions to help you make an informed choice for your application's data storage needs.

For information about storing cryptographic keys, see :ref:`key_storage`.

General recommendations
***********************

For general data storage needs, Nordic Semiconductor recommends starting with the following options:

* NVMC - For applications requiring direct access to non-volatile memory.
* NVS - For most applications requiring simple persistent storage.
  Stores data as values identified by an integer ID.
* ZMS - Recommended for nRF54H and nRF54L Series devices, instead of NVS.

> * RRAMC and ZMS - Recommended for nRF54L Series devices, instead of NVMC and NVS.

Author: OK to keep RRAM/ZMS and NVMC/NVS on the figures?

> * RRAMC and ZMS - Recommended for nRF54L Series devices, instead of NVMC and NVS.

Reviewer 3: nope

.. figure:: /images/data_storage_psa.svg
   :alt: Data storage using PSA Protected Storage

   Data storage using PSA Protected Storage

Key features:

* Authentication and encryption - Data is authenticated and encrypted at rest for confidentiality.
* Isolation - Provides security separation between NSPE and SPE.
* Integrity - Automatic integrity verification.
* Write-once support - Supports write-once flags for immutable data.
* PSA compliance - Compliant with PSA Certified standards.

PSA Protected Storage and TF-M
------------------------------

PSA Protected Storage can be used with or without :ref:`TF-M <ug_tfm>`:

* When used with TF-M, using PSA Protected Storage is mandatory.
  In such case, PSA Protected Storage is an :ref:`Application RoT Service <ug_tfm_architecture_rot_services_application>` that runs in the :ref:`Secure Processing Environment (SPE) <app_boards_spe_nspe>`.
  It is available to both NSPE and SPE, but data stored from one environment is not accessible from the other.
* When used without TF-M, PSA Protected Storage is optional for use in the |NCS|.
  In such case, PSA Protected Storage does not support storing data to external flash.

>   Instead, you can configure your application to encrypt data stored to the external flash, for example using the :ref:`ug_psa_certified_api_overview_crypto`.

Reviewer 4: Protected storage stores data in the internal trusted storage partition which it doesn't support external flash at the moment. In general storing anything in external flash is not supported with the PSA apis at the moment.

When to use PSA Protected Storage
---------------------------------

Consider using PSA Protected Storage in the following cases:

> * You need data authentication and data encryption at rest.

Author: What are the after-effects?

.. _data_storage:
.. _memory_storage:

Data storage in the |NCS|
#########################

.. contents::
   :local:
   :depth: 2

The |NCS| offers multiple solutions for storing data persistently in non-volatile memory.

This page provides an overview of the available storage solutions to help you make an informed choice for your application's data storage needs.

For information about storing cryptographic keys, see :ref:`key_storage`.

General recommendations
***********************

For general data storage needs, Nordic Semiconductor recommends starting with the following options:

* NVMC - For applications requiring direct access to non-volatile memory.
* NVS - For most applications requiring simple persistent storage.
  Stores data as values identified by an 16-bit integer ID.
* ZMS - Recommended for nRF54H and nRF54L Series devices, instead of NVS.

>   Stores data as values identified by an 32-bit integer ID.

Author: This is weird, I remember changing this.

The Non-Volatile Memory Controller (NVMC) provides the most basic method for writing directly to non-volatile memory.
The |NCS| provides the `NVMC driver through nrfx <nrfx NVMC driver_>`_ as a helper for this functionality.

.. note::
   Nordic Semiconductor does not recommend this method for persistent storage, as it places significant responsibility on the developer.
   For simple data storage needs, consider using :ref:`NVS <data_storage_nvs>` or :ref:`ZMS <data_storage_zms>` instead.

.. figure:: /images/data_storage_nvmc.svg
   :alt: Data storage using NVMC

   Data storage using NVMC

When using the NVMC directly, you must manually ensure that:

* Your writes do not conflict with other subsystems.
* You stay within designated memory regions.
* You handle error conditions appropriately.

To use NVMC safely:

1. Create a custom partition for your memory usage.
2. Keep all NVMC writes within this designated region.
3. Coordinate with other subsystems that may use non-volatile memory.

Since NVMC allows writing to any part of non-volatile memory, it can easily overwrite memory used by other subsystems if not used carefully.

.. _data_storage_rramc:

RRAMC
=====

.. note::

>    RRAMC is only available on the `nRF54L Series <nRF54L15 RRAMC_>`_ and should be used with :ref:`ZMS <data_storage_zms>`.

Reviewer 3: Should be used through ZMS rather than directly? Not sure what we're trying to say here.

.. list-table:: RRAMC features
   :widths: auto
   :header-rows: 1

   * - Feature
     - Partitioning
     - Integrity
     - Isolation
     - Encryption
   * - **RRAMC**
     - Yes
     - Yes
     - No
     - No

The Resistive Random Access Memory Controller (RRAMC) provides advanced functionality for managing RRAM memory, including the secure information configuration region (SICR) and user information configuration registers (UICR).

.. figure:: /images/data_storage_rramc.svg
   :alt: Data storage using RRAMC

   Data storage using RRAMC

This option offers the following features:

> * Write without erase - RRAMC allows writing and overwriting without requiring erase operations.

Reviewer 3: Ahem yeah, but this is a feature of the HW. :sweat_smile: I think if we want to talk about RRAMC use directly we should make it similar to NVMC. And again keeping it mind that which of the two you can use simply depends on what device you have.

     - Isolation
     - Encryption
   * - **PSA Protected Storage with TF-M**
     - Yes
     - Yes
     - Yes
     - Yes
   * - **PSA Protected Storage without TF-M**
     - Yes
     - Yes
     - No
     - Yes

The PSA Protected Storage API provides functionality for writing and reading data from non-volatile memory with enhanced security features.
This storage solution is part of the :ref:`ug_psa_certified_api_overview_secstorage`, which in turn is part of the :ref:`Platform Security Architecture (PSA) <ug_psa_certified_api_overview>` framework.

.. figure:: /images/data_storage_psa.svg
   :alt: Data storage using PSA Protected Storage

   Data storage using PSA Protected Storage

Key features:

* Authentication and encryption - Data is authenticated and encrypted at rest for confidentiality.
* Isolation - Provides security separation between NSPE and SPE.
* Integrity - Automatic integrity verification.
* Write-once support - Supports write-once flags for immutable data.
* PSA compliance - Compliant with PSA Certified standards.

PSA Protected Storage and TF-M
------------------------------

PSA Protected Storage can be used with or without :ref:`TF-M <ug_tfm>`:

> * When used with TF-M, PSA Protected Storage is an :ref:`Application RoT Service <ug_tfm_architecture_rot_services_application>` that runs in the :ref:`Secure Processing Environment (SPE) <app_boards_spe_nspe>`.

Reviewer 3: ```suggestion
* When using TF-M, PSA Protected Storage is an :ref:`Application RoT Service <ug_tfm_architecture_rot_services_application>` that runs in the :ref:`Secure Processing Environment (SPE) <app_boards_spe_nspe>`.
```
and same below

  It is available to both NSPE and SPE, but data stored from one environment is not accessible from the other.
* When used without TF-M, PSA Protected Storage does not support storing data to external flash.

>   Instead, at the time of writing to memory, the application will write PSA Protected Storage data to the same location as PS Internal Trusted Storage.

Reviewer 3: When I said `at the time of writing` I meant when this documentation was written. :smile: 
```suggestion
  The data stored through the PSA Protected Storage API will be written to the same location as the PSA Internal Trusted Storage.
```

* Settings integration - Can be used directly or through the :ref:`Settings subsystem <zephyr:settings_api>`.

Configuration
-------------

To use ZMS, enable the :kconfig:option:`CONFIG_ZMS` Kconfig option.

For more information, see also :ref:`zms_memory_storage` in the |NCS| documentation and :ref:`ZMS documentation in Zephyr <zephyr:zms_api>`.

.. _data_storage_filesystems:

File Systems
============

.. list-table:: File Systems features
   :widths: auto
   :header-rows: 1

   * - Feature
     - Partitioning
     - Integrity
     - Isolation
     - Encryption
   * - **File Systems**
     - Yes
     - Yes
     - No
     - No

Zephyr's Virtual Filesystem Switch (VFS) allows applications to mount multiple file systems at different mount points using the :ref:`File System API <zephyr:file_system_api>`.

.. figure:: /images/data_storage_file_system.svg
   :alt: Data storage using file systems

   Data storage using file systems

Two file systems are supported:

* FAT Filesystem - Cross-platform compatible format for data sharing.
* LittleFS - Power loss-resilient filesystem optimized for microcontrollers.

Configuration
-------------

> File systems require storage media configuration (flash, SD card) and the appropriate file system driver:

Author: backend?

> File systems require storage media configuration (flash, SD card) and the appropriate file system driver:

Reviewer 3: No, it really is just file system: https://en.wikipedia.org/wiki/File_system
Maybe you could replace the end by something like `and to enable the appropriate Kconfig options`. Or then just rework the whole phrase/split it into two sentences.

The Settings subsystem does not write to non-volatile memory directly.
Instead, it uses other subsystems as backends, with :ref:`data_storage_nvs` being the recommended choice for most devices and :ref:`data_storage_zms` for nRF54L Series and nRF54H20 devices.

You can use the ``settings_load_one()`` function to load only the value defined by the tree name given in the arguments.
This is a very fast way to get a value from the storage if you know its tree name.
It is particularly suitable for use with ZMS because it does not need to load all settings before getting the right one.

Settings are used by various subsystems, including :ref:`Bluetooth host stack <zephyr:bluetooth-persistent-storage>` and :ref:`Matter <ug_matter>`.

Configuration
-------------

To use Settings, enable the :kconfig:option:`CONFIG_SETTINGS` Kconfig option.

Depending on the backend you want to use, enable the appropriate Kconfig options:

* For the NVS backend:

  * :kconfig:option:`CONFIG_SETTINGS_NVS`
  * :kconfig:option:`CONFIG_NVS`
  * :kconfig:option:`CONFIG_FLASH_MAP`

* For the ZMS backend:

  * :kconfig:option:`CONFIG_SETTINGS_ZMS`
  * :kconfig:option:`CONFIG_ZMS`

> * For the File System backend:

Reviewer 3: Backend here is correct because we're talking about Settings backends.

  * :kconfig:option:`CONFIG_SETTINGS_FILE`

>   * :kconfig:option:`CONFIG_FILE_SYSTEM` - with the appropriate file system backend enabled, as mentioned in :ref:`data_storage_filesystems`

Reviewer 3: ```suggestion
  * :kconfig:option:`CONFIG_FILE_SYSTEM` - with the appropriate file system enabled, as mentioned in :ref:`data_storage_filesystems`
```
